### PR TITLE
[Feature]Alter lake table add partition

### PR DESCRIPTION
**Pull request by abc982627271**
## What type of PR is this：
- [ ] bug
- [x] feature
- [ ] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：

Fixes #

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
Can't set status; build succeeded.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 28 / 79 (35.44%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/persist/AddPartitionsInfoV2.java|0|8|00.00%|[20, 21, 22, 26, 27, 30, 31, 35]|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|0|7|00.00%|[225, 226, 227, 228, 229, 1101, 1102]|
|:large_blue_circle:|com/starrocks/persist/PartitionPersistInfoV2.java|0|4|00.00%|[52, 53, 54, 112]|
|:large_blue_circle:|com/starrocks/persist/RangePartitionPersistInfo.java|0|1|00.00%|[36]|
|:large_blue_circle:|com/starrocks/journal/JournalEntity.java|0|3|00.00%|[236, 237, 238]|
|:large_blue_circle:|com/starrocks/catalog/RangePartitionInfo.java|1|8|12.50%|[249, 250, 251, 252, 253, 254, 255]|
|:large_blue_circle:|com/starrocks/catalog/CatalogUtils.java|1|2|50.00%|[37]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|26|46|56.52%|[1085, 1086, 1087, 1088, 1089, 1090, 1091, 1092, 1093, 1111, 1112, 1113, 1114, 1115, 1116, 1118, 1119, 1133, 1134, 1135]|

File: fe/fe-core/src/main/java/com/starrocks/catalog/CatalogUtils.java [hunk 1]
```diff
@@ -32,9 +32,9 @@ public static void checkTableExist(Database db, String tableName) throws DdlExce
     }
 
     // check table type is OLAP
-    public static void checkTableTypeOLAP(Database db, Table table) throws DdlException {
-        if (table.getType() != Table.TableType.OLAP && table.getType() != Table.TableType.MATERIALIZED_VIEW) {
-            throw new DdlException("Table[" + table.getName() + "] is not OLAP table");
+    public static void checkNativeTable(Database db, Table table) throws DdlException {
+        if (!table.isOlapOrLakeTable() && table.getType() != Table.TableType.MATERIALIZED_VIEW) {
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/catalog/CatalogUtils.java:36]: !table.isNativeTable()

File: fe/fe-core/src/main/java/com/starrocks/persist/OperationType.java [hunk 2]
```diff
@@ -256,6 +256,10 @@ public class OperationType {
     // only used in list partition currently
     public static final short OP_ADD_PARTITION_V2 = 10241;
 
+    // new operator for add partition 10261 ~ 10280
+    // only used in lake table currently
+    public static final short OP_ADD_PARTITIONS_V2 = 10261;
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/persist/OperationType.java:261]:
```suggestion
    public static final short OP_ADD_PARTITIONS_V2 = 10242;
```

File: fe/fe-core/src/main/java/com/starrocks/server/LocalMetastore.java [hunk 3]
```diff
@@ -1071,20 +1073,32 @@ private void addRangePartitionLog(Database db, OlapTable olapTable, List<Partiti
         // Forward compatible with previous log formats
         // Version 1.15 is compatible if users only use single-partition syntax.
         // Otherwise, the followers will be crash when reading the new log
+        List<PartitionPersistInfoV2> partitionInfoListV2 = Lists.newArrayListWithCapacity(partitionLen);
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/server/LocalMetastore.java:1076]:
```suggestion
        List<PartitionPersistInfoV2> partitionInfoV2List = Lists.newArrayListWithCapacity(partitionLen);
```

File: fe/fe-core/src/main/java/com/starrocks/server/LocalMetastore.java [hunk 4]
```diff
@@ -1093,19 +1107,35 @@ private void addRangePartitionLog(Database db, OlapTable olapTable, List<Partiti
...
-                                    partitionInfo.getIsInMemory(partition.getId()),
-                                    isTempPartition);
-                    partitionInfoList.add(info);
+                    if (olapTable.isLakeTable()) {
+                        PartitionPersistInfoV2 info = new RangePartitionPersistInfo(db.getId(), olapTable.getId(), 
+                                partition, partitionDescs.get(i).getPartitionDataProperty(), 
+                                partitionInfo.getReplicationNum(partition.getId()),
+                                partitionInfo.getIsInMemory(partition.getId()), isTempPartition,
+                                ((RangePartitionInfo) partitionInfo).getRange(partition.getId()),
+                                ((SingleRangePartitionDesc) partitionDescs.get(0)).getStorageInfo());
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/server/LocalMetastore.java:1116]:
```suggestion
                                ((SingleRangePartitionDesc) partitionDescs.get(i)).getStorageInfo());
```

File: fe/fe-core/src/main/java/com/starrocks/server/LocalMetastore.java [hunk 5]
```diff
@@ -2549,7 +2582,7 @@ private void createLakeTablets(LakeTable table, long partitionId, MaterializedIn
         }
 
         PartitionInfo partitionInfo = table.getPartitionInfo();
-        StorageInfo partitionStorageInfo = partitionInfo.getStorageInfo(partitionId);
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/server/LocalMetastore.java:2552]: Each partition has his own StorageInfo that only include cache info.
Here we should use partition cache info.

**Comment by wyb** [comment]
Should clear tablet and shard when add partition failed.

**Comment by wyb** [comment]
Add more UT.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 53 / 87 (60.92%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/journal/JournalEntity.java|0|3|00.00%|[236, 237, 238]|
|:large_blue_circle:|com/starrocks/catalog/RangePartitionInfo.java|1|8|12.50%|[249, 250, 251, 252, 253, 254, 255]|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|2|7|28.57%|[225, 226, 227, 228, 229]|
|:large_blue_circle:|com/starrocks/catalog/CatalogRecycleBin.java|2|5|40.00%|[933, 934, 935]|
|:large_blue_circle:|com/starrocks/catalog/CatalogUtils.java|1|2|50.00%|[37]|
|:large_blue_circle:|com/starrocks/persist/AddPartitionsInfoV2.java|5|8|62.50%|[30, 31, 35]|
|:large_blue_circle:|com/starrocks/persist/PartitionPersistInfoV2.java|3|4|75.00%|[112]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|38|49|77.55%|[1117, 1118, 1119, 1120, 1121, 1122, 1124, 1125, 1139, 1140, 1141]|
|:large_blue_circle:|com/starrocks/persist/RangePartitionPersistInfo.java|1|1|100.00%|[]|

**Comment by abc982627271** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 53 / 87 (60.92%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/journal/JournalEntity.java|0|3|00.00%|[236, 237, 238]|
|:large_blue_circle:|com/starrocks/catalog/RangePartitionInfo.java|1|8|12.50%|[249, 250, 251, 252, 253, 254, 255]|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|2|7|28.57%|[225, 226, 227, 228, 229]|
|:large_blue_circle:|com/starrocks/catalog/CatalogRecycleBin.java|2|5|40.00%|[933, 934, 935]|
|:large_blue_circle:|com/starrocks/catalog/CatalogUtils.java|1|2|50.00%|[37]|
|:large_blue_circle:|com/starrocks/persist/AddPartitionsInfoV2.java|5|8|62.50%|[30, 31, 35]|
|:large_blue_circle:|com/starrocks/persist/PartitionPersistInfoV2.java|3|4|75.00%|[112]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|38|49|77.55%|[1117, 1118, 1119, 1120, 1121, 1122, 1124, 1125, 1139, 1140, 1141]|
|:large_blue_circle:|com/starrocks/persist/RangePartitionPersistInfo.java|1|1|100.00%|[]|

**Comment by abc982627271** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 53 / 87 (60.92%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/journal/JournalEntity.java|0|3|00.00%|[236, 237, 238]|
|:large_blue_circle:|com/starrocks/catalog/RangePartitionInfo.java|1|8|12.50%|[249, 250, 251, 252, 253, 254, 255]|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|2|7|28.57%|[225, 226, 227, 228, 229]|
|:large_blue_circle:|com/starrocks/catalog/CatalogRecycleBin.java|2|5|40.00%|[933, 934, 935]|
|:large_blue_circle:|com/starrocks/catalog/CatalogUtils.java|1|2|50.00%|[37]|
|:large_blue_circle:|com/starrocks/persist/AddPartitionsInfoV2.java|5|8|62.50%|[30, 31, 35]|
|:large_blue_circle:|com/starrocks/persist/PartitionPersistInfoV2.java|3|4|75.00%|[112]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|38|49|77.55%|[1117, 1118, 1119, 1120, 1121, 1122, 1124, 1125, 1139, 1140, 1141]|
|:large_blue_circle:|com/starrocks/persist/RangePartitionPersistInfo.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 64 / 87 (73.56%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/journal/JournalEntity.java|0|3|00.00%|[236, 237, 238]|
|:large_blue_circle:|com/starrocks/catalog/RangePartitionInfo.java|1|8|12.50%|[249, 250, 251, 252, 253, 254, 255]|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|2|7|28.57%|[225, 226, 227, 228, 229]|
|:large_blue_circle:|com/starrocks/catalog/CatalogRecycleBin.java|2|5|40.00%|[933, 934, 935]|
|:large_blue_circle:|com/starrocks/catalog/CatalogUtils.java|1|2|50.00%|[37]|
|:large_blue_circle:|com/starrocks/persist/AddPartitionsInfoV2.java|5|8|62.50%|[30, 31, 35]|
|:large_blue_circle:|com/starrocks/persist/PartitionPersistInfoV2.java|3|4|75.00%|[112]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|49|49|100.00%|[]|
|:large_blue_circle:|com/starrocks/persist/RangePartitionPersistInfo.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 72 / 87 (82.76%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/journal/JournalEntity.java|0|3|00.00%|[236, 237, 238]|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|2|7|28.57%|[225, 226, 227, 228, 229]|
|:large_blue_circle:|com/starrocks/catalog/CatalogRecycleBin.java|2|5|40.00%|[933, 934, 935]|
|:large_blue_circle:|com/starrocks/catalog/CatalogUtils.java|1|2|50.00%|[37]|
|:large_blue_circle:|com/starrocks/persist/AddPartitionsInfoV2.java|5|8|62.50%|[30, 31, 35]|
|:large_blue_circle:|com/starrocks/catalog/RangePartitionInfo.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|49|49|100.00%|[]|
|:large_blue_circle:|com/starrocks/persist/PartitionPersistInfoV2.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/persist/RangePartitionPersistInfo.java|1|1|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/persist/OperationType.java [hunk 6]
```diff
@@ -256,6 +256,10 @@ public class OperationType {
     // only used in list partition currently
     public static final short OP_ADD_PARTITION_V2 = 10241;
 
+    // new operator for add partition 10261 ~ 10280
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/persist/OperationType.java:259]: Remove this comment.

File: fe/fe-core/src/main/java/com/starrocks/server/LocalMetastore.java [hunk 7]
```diff
@@ -993,6 +995,12 @@ private List<Partition> createPartitionList(Database db, OlapTable copiedTable,
             copiedTable.getPartitionInfo()
                     .setReplicationNum(partitionId, partitionDesc.getReplicationNum());
             copiedTable.getPartitionInfo().setIsInMemory(partitionId, partitionDesc.isInMemory());
+            // for debug
+            // LOG.info("partitionDesc type is {}", partitionDesc.getClass().getName());
+            if (partitionDesc instanceof SingleRangePartitionDesc) {
+                copiedTable.getPartitionInfo().setStorageInfo(partitionId,
+                        ((SingleRangePartitionDesc) partitionDesc).getStorageInfo());
+            }
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/server/LocalMetastore.java:1003]: Better to abstract a `getStorageInfo` method in `PartitionDesc`.

File: fe/fe-core/src/main/java/com/starrocks/server/LocalMetastore.java [hunk 8]
```diff
@@ -1265,6 +1301,8 @@ private void addPartitions(Database db, String tableName, List<PartitionDesc> pa
             }
         } catch (DdlException e) {
             cleanTabletIdSetForAll(tabletIdSetForAll);
+            // lake table need to delete shards
+            stateMgr.getShardManager().getShardDeleter().addUnusedShardId(tabletIdSetForAll);
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/server/LocalMetastore.java:1305]: Check lake table and put it in `cleanTabletIdSetForAll`.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 71 / 85 (83.53%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/analysis/PartitionDesc.java|0|1|00.00%|[91]|
|:large_blue_circle:|com/starrocks/journal/JournalEntity.java|0|3|00.00%|[236, 237, 238]|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|2|7|28.57%|[225, 226, 227, 228, 229]|
|:large_blue_circle:|com/starrocks/catalog/CatalogUtils.java|1|2|50.00%|[37]|
|:large_blue_circle:|com/starrocks/persist/AddPartitionsInfoV2.java|5|8|62.50%|[30, 31, 35]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|48|49|97.96%|[1209]|
|:large_blue_circle:|com/starrocks/catalog/RangePartitionInfo.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/analysis/SingleItemListPartitionDesc.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/persist/PartitionPersistInfoV2.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/persist/RangePartitionPersistInfo.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/analysis/MultiItemListPartitionDesc.java|1|1|100.00%|[]|

**Review by gengjun-git — APPROVED** [review]

**Review by wyb — APPROVED** [review]

**Comment by wyb** [comment]
run starrocks_admit_test

**Comment by wyb** [comment]
run starrocks_admit_test

**Pull request merged**